### ISSUE TITLE
Whitespace tokenizer for training BERT from scratch

### ISSUE THREAD

**Issue by aqibsaeed** [huggingface/tokenizers#232]
Is there any example of using a whitespace tonkenizer (that splits text based only on whitespaces) for [training BERT](https://huggingface.co/blog/how-to-train) ?

**Comment by julien-c** [comment]
See also https://github.com/huggingface/transformers/issues/3774

**Comment by parmarsuraj99** [comment]
`transformers.BertTokenizerFast(vocab_file, do_lower_case=True, unk_token='[UNK]', sep_token='[SEP]', pad_token='[PAD]', cls_token='[CLS]', mask_token='[MASK]', clean_text=True, tokenize_chinese_chars=True, strip_accents=True, wordpieces_prefix='##', **kwargs)`

Maybe you can use this. By changing `wordpieces_prefix`to `" "`

**Comment by n1t0** [comment]
Hi @aqibsaeed, please read my answer in another issue that should give you some direction on how to do this. Here is the answer in question: https://github.com/huggingface/tokenizers/issues/243#issuecomment-617860020

There is a `WhitespaceSplit` `PreTokenizer` that does split on whitespace only.

The default `PreTokenizer` we use in `BertWordPieceTokenizer` actually splits on whitespace, and also punctuation.

**Comment by snowzjy** [comment]
I also need this very mush

**Comment by snowzjy** [comment]
> Hi @aqibsaeed, please read my answer in another issue that should give you some direction on how to do this. Here is the answer in question: [#243 (comment)](https://github.com/huggingface/tokenizers/issues/243#issuecomment-617860020)
> 
> There is a `WhitespaceSplit` `PreTokenizer` that does split on whitespace only.
> 
> The default `PreTokenizer` we use in `BertWordPieceTokenizer` actually splits on whitespace, and also punctuation.

Useless, there is not WhiteSpaceTrainer, only have BpeTrainer and WordPieceTrainer, so I can't train a whitespace tokenizer by python code.

**Comment by aqibsaeed** [comment]
This does not really work! The tokenizer results in weird tokenization of the text. It is super strange that there is no simple example available to achieve this.

**Comment by aqibsaeed** [comment]
`from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.normalizers import Lowercase
from tokenizers.pre_tokenizers import CharDelimiterSplit

# We build our custom tokenizer:
tokenizer = Tokenizer(BPE()) 
tokenizer.normalizer = Lowercase()
tokenizer.pre_tokenizer = CharDelimiterSplit('_')

# We can train this tokenizer by giving it a list of path to text files:
trainer = BpeTrainer(special_tokens=[ "[UNK]" ])
tokenizer.train(trainer, ["1.txt", "2.txt", "3.txt"])

# And simply use it
tokenizer.encode('Hello_there!')`

If I use the exact same code that you posted in another thread `tokenizer.encode('Hello_there!')` results in 6 tokens! whereas it should results in two tokens with `CharDelimiterSplit('_')`. Isn't it?

**Comment by moranbel** [comment]
use `WordLevel` as Tokenizer.
see here- https://huggingface.co/docs/tokenizers/python/latest/components.html
